libavcodec/vaapi_encode_vp9.c:set a defualt value for log2_tile_columns

zhangyuankun-star opened the pull request:
Maximum width of a tile in units of superblocks is 64,
So the number of tile columns is related to the width of the picture.
We should set a defualt value for log2_tile_columns
to to avoid errors in video coding with width over 4096.

Signed-off-by: Zhang yuankun <yuankunx.zhang@intel.com>

zhangyuankun-star commented:
@xhaihao @xuguangxin , Can you review this for me ?

xuguangxin commented on libavcodec/vaapi_encode_vp9.c:
```diff
@@ -82,10 +83,16 @@ static int vaapi_encode_vp9_init_picture_params(AVCodecContext *avctx,
 
+    // Maximum width of a tile in units of superblocks is 64
+    // So the number of tile columns is related to the width of the picture.
+    numTileColumns = (vpic->frame_width_src + VP9_MAX_TILE_WIDTH - 1) / VP9_MAX_TILE_WIDTH;
```
please change to num_tile_columns.

xuguangxin commented on libavcodec/vaapi_encode_vp9.c:
```diff
@@ -82,10 +83,16 @@ static int vaapi_encode_vp9_init_picture_params(AVCodecContext *avctx,
     vpic->reconstructed_frame = pic->recon_surface;
     vpic->coded_buf = pic->output_buffer;
 
+    // Maximum width of a tile in units of superblocks is 64
```
change 64 to MAX_TILE_WIDTH_B64(64)

xuguangxin reviewed (changes requested):
Did you checked tile width < MIN_TILE_WIDTH_B64(4) super block?

xuguangxin commented on libavcodec/vaapi_encode_vp9.c:
```diff
@@ -82,10 +83,17 @@ static int vaapi_encode_vp9_init_picture_params(AVCodecContext *avctx,
+    // So the number of tile columns is related to the width of the picture.
+    // We set the minimum possible number for num_tile_columns as defualt value.
+    num_tile_columns = (vpic->frame_width_src + VP9_MAX_TILE_WIDTH - 1) / VP9_MAX_TILE_WIDTH;
+    vpic->log2_tile_columns = av_log2(num_tile_columns);
```
this may not right. if you have a 9k video,  num_tile_columns will be 3. vpic->log2_tile_columns will be 1.   each tile columns will large than 4k

xuguangxin reviewed (changes requested):
Please refer to other ffmpeg commit log and change commit message too.

xuguangxin reviewed (changes requested):
How about this:
```
avcodec/vaapi_encode_vp9: fix > 4k encode fail issue

This patch will fix following command:
xxxxx

Max width of a vp9 tile is 4096.  If we the source frame > 4096, we need split to multiple tiles.
```
also did you checked MAX_TILE_AREA? what's happened if you encode a 9k x 4096 video?

xuguangxin commented:
Max width of a vp9 tile is 4096.  If ~~we~~ the source frame > 4096, we need split to multiple tiles.
Sorry, we need remove the we.

could we use a simplified pipeline like
ffmpeg -hwaccel vaapi -hwaccel_output_format vaapi -i input.264 -vf 'scale_vaapi=w=7680:h=4096' -c:v vp9_vaapi output.ivf

zhangyuankun-star committed:
avcodec/vaapi_encode_vp9: fix > 4k encode fail issue

This patch will fix following command:
ffmpeg -hwaccel vaapi -hwaccel_output_format vaapi -i input.264 \
-vf 'scale_vaapi=w=7680:h=4096' -c:v vp9_vaapi output.ivf

Max width of a vp9 tile is 4096. If the source frame > 4096, we need split to multiple tiles.

Signed-off-by: Zhang yuankun <yuankunx.zhang@intel.com>

zhangyuankun-star commented:
I have send this patch to ffmpeg community, so I close this PR.
link:https://patchwork.ffmpeg.org/project/ffmpeg/patch/20210412014724.18563-1-yuankunx.zhang@intel.com/
